edx_arch_experiments/scripts/republish_failed_events.py

```diff
@@ -0,0 +1,79 @@
+
+This is created as a script instead of a management command because it's meant to be used as a one-off and not to
+require pip installing this package into anything else to run. However, since edx-event-bus-kafka does expect certain
+settings, the script must be run in an environment with DJANGO_SETTINGS_MODULE.
+
+To run:
+tox -e scripts -- python edx_arch_experiments/scripts/republish_failed_events.py
+ --filename /Users/rgraber/oneoffs/failed_events.csv
+"""
+
+import csv
+import json
+import sys
+from ast import literal_eval
+
+import click
+from edx_event_bus_kafka.internal.producer import create_producer
+from openedx_events.tooling import EventsMetadata, OpenEdxPublicSignal, load_all_signals
+
+
+@click.command()
+@click.option('--filename', type=click.Path(exists=True))
+def read_and_send_events(filename):
+    load_all_signals()
+    producer = create_producer()
+    try:
+        log_columns = ['initial_topic', 'event_type', 'event_data_as_json', 'event_key_field', 'event_metadata_as_json']
+        with open(filename) as log_file:
+            reader = csv.DictReader(log_file)
+            # Make sure csv contains all necessary columns for republishing
+            if not all(column in reader.fieldnames for column in log_columns):
+                print(f'Missing required columns {set(log_columns).difference(set(reader.fieldnames))}. Cannot'
+                      f' republish events.')
+                sys.exit(1)
+            ids = set()
+            for row in reader:
+                # We log everything using __repr__, so strings get quotes around them and "None" gets
+                # written literally. Use literal_eval to go from "None" to None and remove the extraneous quotes
+                # from the logs
+                empties = [key for key, value in row.items() if key in log_columns and literal_eval(value) is None]
+                # If any row is missing data, stop processing the whole file to avoid sending events out of order
+                if len(empties) > 0:
+                    print(f'Missing required fields in row {reader.line_num}: {empties}. Will not continue publishing.')
+                    sys.exit(1)
+
+                topic = literal_eval(row['initial_topic'])
+                event_type = literal_eval(row['event_type'])
+                event_data = json.loads(literal_eval(row['event_data_as_json']))
+                event_key_field = literal_eval(row['event_key_field'])
+                metadata = EventsMetadata.from_json(literal_eval(row['event_metadata_as_json']))
+                signal = OpenEdxPublicSignal.get_signal_by_type(event_type)
+                if metadata.id in ids:
+                    print(f"Skipping duplicate id {metadata.id}")
+                    continue
+                ids.add(metadata.id)
+
+                producer.send(signal=signal, event_data=event_data, event_key_field=event_key_field, topic=topic,
+                              event_metadata=metadata)
```

Comment:
I think we should print something after every event that is successfully sent so that if the script dies partway through (parsing errors, network trouble, etc.), we can figure out where to resume from. Currently we only get that information when there's `None` in a required field.

Reply:
It also occurs to me now that including the CSV line number might be useful.